main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -212,8 +212,12 @@ object Inliner {
       /// Both code size and runtime are reduced
       if (isDeadAndPure(occur, exp1.purity)) {
         visitExp(exp2, subst0)
-      } else {
+      } else if (occur == Dead && occur != DontInline) {
```

Comment:
I would prefer a helper method named isDeadAndImpure right next to isDeadAndPure.